### ISSUE TITLE
Device tree of Ruijie RG-EW3200GX PRO is out of date

### ISSUE THREAD

**Issue by asterwyx** [openwrt/openwrt#15924]
### Describe the bug

Firmware compiled from the master branch for Ruijie RG-EW3200GX PRO is not bootable cause the device tree is out of date. I've dumped my device tree and acquired the updated device tree. I'm not sure if I could contribute to this repo cause this is a version-specific problem. How do you manage multiple versions of device tree?

### OpenWrt version

master

### OpenWrt release

master

### OpenWrt target/subtarget

mediatek/mt7622

### Device

Ruijie RG-EW3200GX PRO

### Image kind

Self-built image

### Terms

- [X] I am reporting an issue for OpenWrt, not an unsupported fork.

**Comment by RolandoMagico** [comment]
Hi @asterwyx ,
can you provide more information? What are the differences in the device tree? Are there different hardware revisions of this router which require different device trees?

**Comment by asterwyx** [comment]
Yes, there are different hardware revisions of this router I think. In my fork of [openwrt](https://github.com/asterwyx/openwrt) branch [rg](https://github.com/asterwyx/openwrt/tree/rg) you can find my patch. It works for me now.

**Comment by RolandoMagico** [comment]
Do you know which hardware version you're using?
There are already two variants supported in OpenWrt:
- The initial one using the memory layout here: https://github.com/asterwyx/openwrt/blob/rg/target/linux/mediatek/dts/mt7622-ruijie-rg-ew3200gx-pro.dts
- A model variant "E5" which uses a diferent memory layout: https://github.com/asterwyx/openwrt/blob/rg/target/linux/mediatek/dts/mt7622-reyee-ax3200-e5.dts

So if you know your hardware revision, you can create a pull request similar to the one here:
https://github.com/openwrt/openwrt/commit/7dbac3433fef1af0f76a515e962db450c3407c19

**Comment by Shine-** [comment]
I got my hands on one of these devices and can confirm this.

Mine is using the flash layout of the model called "reyee AX3200 E5" in OpenWrt, while everywhere on the box, and also on the bottom of the device, it says "Ruijie/Reyee RG-EW3200GX PRO". In addition, it says hardware version 1.10 on the box, on the device bottom and stock firmware. The stock firmware version is 1.219 and displays the device model as EW3200GX-PRO.

I assume this partitioning change happened with hardware version 1.10, but it could of course also be possible that a specific firmware update auto-repartitioned the device on install.

Perhaps someone else could confirm/deny whether this change in partitioning happened at a specific firmware version, or hardware version? There's currently no earlier firmware than 1.313 (December 2024) available on the vendor web site.

As a word of warning, the partition specification in the "reyee AX3200 E5" DTS was also incorrect for years. It's since been fixed in SNAPSHOT, but it's still wrong in 24.10 and earlier, so beware when trying to install STABLE on this device. You could also backport f4ba980f154fabc63a7164309543312b320cbfd7 to fix.

**TL;DR**
You can either use current SNAPSHOT build for "reyee AX3200 E5", or apply this patch for proper support (build and run tested with 24.10.4 **only**):
<details><summary>Ruijie RG-EW3200GX PRO V1.10 device support</summary>

```diff
diff --git a/package/boot/uboot-envtools/files/mediatek_mt7622 b/package/boot/uboot-envtools/files/mediatek_mt7622
index fd40b664e4..e485d748bf 100644
--- a/package/boot/uboot-envtools/files/mediatek_mt7622
+++ b/package/boot/uboot-envtools/files/mediatek_mt7622
@@ -48,6 +48,10 @@ linksys,e8450-ubi)
 buffalo,wsr-2533dhp2)
 	ubootenv_add_uci_config "/dev/mtd3" "0x0" "0x1000" "0x20000"
 	;;
+reyee,ax3200-e5|\
+ruijie,rg-ew3200gx-pro-v110)
+	ubootenv_add_uci_config "/dev/mtd3" "0x0" "0x10000" "0x10000"
+	;;
 ruijie,rg-ew3200gx-pro)
 	ubootenv_add_uci_config "/dev/mtd3" "0x0" "0x20000" "0x20000"
 	;;
diff --git a/target/linux/mediatek/dts/mt7622-ruijie-rg-ew3200gx-pro-v110.dts b/target/linux/mediatek/dts/mt7622-ruijie-rg-ew3200gx-pro-v110.dts
new file mode 100644
index 0000000000..4dbc688e2a
--- /dev/null
+++ b/target/linux/mediatek/dts/mt7622-ruijie-rg-ew3200gx-pro-v110.dts
@@ -0,0 +1,96 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+/dts-v1/;
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/gpio/gpio.h>
+
+#include "mt7622-ruijie-rg-ew3200.dtsi"
+
+/ {
+	model = "Ruijie RG-EW3200GX PRO V1.10";
+	compatible = "ruijie,rg-ew3200gx-pro-v110", "mediatek,mt7622";
+};
+
+&nor_flash {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi_nor_pins>;
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <50000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "Preloader";
+				reg = <0x0 0x20000>;
+				read-only;
+			};
+
+			partition@20000 {
+				label = "ATF";
+				reg = <0x20000 0x10000>;
+				read-only;
+			};
+
+			partition@30000 {
+				label = "u-boot";
+				reg = <0x30000 0x50000>;
+				read-only;
+			};
+
+			partition@80000 {
+				label = "u-boot-env";
+				reg = <0x80000 0x10000>;
+			};
+
+			factory: partition@90000 {
+				label = "Factory";
+				reg = <0x90000 0x40000>;
+				read-only;
+			};
+
+			partition@d0000 {
+				label = "product_info";
+				reg = <0xd0000 0x10000>;
+				read-only;
+			};
+
+			partition@e0000 {
+				label = "kdump";
+				reg = <0xe0000 0x10000>;
+				read-only;
+			};
+
+			partition@f0000 {
+				compatible = "denx,fit";
+				label = "firmware";
+				reg = <0xf0000 0xf10000>;
+			};
+		};
+	};
+};
+
+&wmac {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&epa_elna_pins>;
+	mediatek,mtd-eeprom = <&factory 0x0>;
+};
+
+&slot0 {
+	wifi@0,0 {
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x5000>;
+		ieee80211-freq-limit = <5000000 6000000>;
+		mediatek,disable-radar-background;
+	};
+};
diff --git a/target/linux/mediatek/image/mt7622.mk b/target/linux/mediatek/image/mt7622.mk
index faf1b20e4a..aeb92ec2c9 100644
--- a/target/linux/mediatek/image/mt7622.mk
+++ b/target/linux/mediatek/image/mt7622.mk
@@ -318,12 +318,23 @@ define Device/ruijie_rg-ew3200gx-pro
 endef
 TARGET_DEVICES += ruijie_rg-ew3200gx-pro
 
+define Device/ruijie_rg-ew3200gx-pro-v110
+  DEVICE_VENDOR := Ruijie
+  DEVICE_MODEL := RG-EW3200GX PRO V1.10
+  DEVICE_DTS := mt7622-ruijie-rg-ew3200gx-pro-v110
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_PACKAGES := kmod-mt7915-firmware
+  SUPPORTED_DEVICES += reyee,ax3200-e5
+endef
+TARGET_DEVICES += ruijie_rg-ew3200gx-pro-v110
+
 define Device/reyee_ax3200-e5
   DEVICE_VENDOR := reyee
   DEVICE_MODEL := AX3200 E5
   DEVICE_DTS := mt7622-reyee-ax3200-e5
   DEVICE_DTS_DIR := ../dts
   DEVICE_PACKAGES := kmod-mt7915-firmware
+  SUPPORTED_DEVICES += ruijie,rg-ew3200gx-pro-v110
 endef
 TARGET_DEVICES += reyee_ax3200-e5
 
```
Credits:

- ce8a33b021a703b1933f627835f6cdc88c0ecdd3, d15f9b9043eac41ecfa1aaf1d65dca460b3c7fb4 and 7dbac3433fef1af0f76a515e962db450c3407c19 for initial support
- f4ba980f154fabc63a7164309543312b320cbfd7 for DTS fix
- my additional work in creating this patch is Public Domain, no credit needed or wanted

</details>
